R/parameter_tuning.R

```diff
@@ -0,0 +1,218 @@
+  if (!is.vector(smoothing_bandwidth) || length(smoothing_bandwidth) == 1) {
+    smoothing_bandwidth = rep(smoothing_bandwidth, length(attributes_to_visualize))
+  }
+
+  # Visualize each attribute
+  for (i in seq_along(attributes_to_visualize)) {
+    # Create a data frame containing the attribute in question and the metric values
+    attribute = attributes_to_visualize[i]
+    df = data.frame(
+      x = as.numeric(parameter_df[, attribute]),
+      y = as.numeric(metric_values)
+    )
+    # Create a smooth trend line using a smoothing kernel
+    smooth_trend_line = as.data.frame(
+      ksmooth(
+        df$x,
+        df$y,
+        kernel = "normal",
+        bandwidth = smoothing_bandwidth[[i]],
+        n.points = 100
+      )
+    )
+
+    # Plot both the scatter plot and the smooth trend line
+    p = ggplot2::ggplot(smooth_trend_line, ggplot2::aes(x = x, y = y)) +
+      ggplot2::geom_line() +
+      ggplot2::geom_point(data = df) +
+      ggplot2::labs(x = attribute, y = "metric value") +
+      ggplot2::ggtitle(title)
+    print(p)
+  }
+}
+
+test_tree_registration = function(
+    map_params,
+    registration_methods,
+    per_method_registration_arguments = NULL,
+    registration_method_names = NULL,
+    n_random_samples = NULL) {
+  #' Test the different tree registration algorithms in a variety of situations
+  #'
+  #' @param map_params Named list. The names correspond to arguments to simulate_tree_maps. The values
+  #' correspond to potential values of this argument to be tried
+  #' @param registration_methods List of registration algorithms
+  #' @param per_method_registration_arguments A list of named lists, one for each registration method.
+  #' The arguments in each list will be passed to the corresponding function
+  #' @param registration_method_names The human-readable names for each method
+  #' @param n_random_samples The number of random samples to draw from the grid of paramters
+  #'
+  #' @returns Named list with the following fields
+  #'   per_method_results: Named list, where each name is from registration_method_names. The value is a data
+  #'     frame of results from the experiments using that method.
+  #'   all_map_param_configurations: Dataframe of paramters used in experiments, one row per trial
+
+  # Get the names of all arguments of the function and their default values
+  simulate_tree_maps_default_args = formals(simulate_tree_maps)
+
+  # Build the list of all provided values while using defaults for unspecified values
+  # Create an empty list
+  all_map_params = list()
+  # Iterate over the function arguments
+  for (name in names(simulate_tree_maps_default_args)) {
+    if (name %in% names(map_params)) {
+      # If the name is in the named list of parameters use that value
+      all_map_params[name] = map_params[name]
+    } else {
+      # Else, use the default value
+      all_map_params[name] = simulate_tree_maps_default_args[name]
+    }
```

Comment:
Yeah, it seems effective. The one issue I had with it involved a negative default being interpreted in a strange way I couldn't understand. I think it may have been split into a negative sign and the value or something.